cpp/src/cylon/serialize/table_serialize.cpp

```diff
@@ -0,0 +1,354 @@
+        auto buf,
+        arrow::internal::CopyBitmap(pool, data.buffers[buf_idx]->data(), data.offset, data.length))
+    data_buffers->push_back(buf->data());
+    bitmaps_with_offset->push_back(std::move(buf));
+  }
+
+  return Status::OK();
+}
+
+Status CollectDataBuffer(const arrow::ArrayData &data, std::vector<int32_t> *buffer_sizes,
+                         std::vector<const uint8_t *> *data_buffers,
+                         arrow::BufferVector *bitmaps_with_offset,
+                         arrow::MemoryPool *pool) {
+  const auto &type = data.type;
+  if (type->id() == arrow::Type::BOOL) {
+    return CollectBitmapInfo<1>(data, buffer_sizes, data_buffers, bitmaps_with_offset, pool);
+  }
+
+  if (arrow::is_fixed_width(type->id())) {
+    int byte_width = std::static_pointer_cast<arrow::FixedWidthType>(type)->bit_width() / CHAR_BIT;
+    buffer_sizes->push_back(byte_width * (int) data.length);
+    data_buffers->push_back(data.buffers[1]->data() + data.offset * byte_width);
+    return Status::OK();
+  }
+
+  if (arrow::is_binary_like(type->id())) {
+    int start_offset = data.GetValues<int32_t>(1)[0];
+    int end_offset = data.GetValues<int32_t>(1)[data.length];
+    buffer_sizes->push_back(end_offset - start_offset);
+    data_buffers->push_back(data.buffers[2]->data() + start_offset);
+    return Status::OK();
+  }
+
+  if (arrow::is_large_binary_like(type->id())) {
+    int start_offset = (int) data.GetValues<int64_t>(1)[0];
+    int end_offset = (int) data.GetValues<int64_t>(1)[data.length];
+    buffer_sizes->push_back(end_offset - start_offset);
+    data_buffers->push_back(data.buffers[2]->data() + start_offset);
+    return Status::OK();
+  }
+
+  return {Code::Invalid, "unsupported data type for serialization " + type->ToString()};
+}
+
+Status CollectOffsetBuffer(const arrow::ArrayData &data, std::vector<int32_t> *buffer_sizes,
+                           std::vector<const uint8_t *> *data_buffers,
+                           arrow::BufferVector *extra_buffers, arrow::MemoryPool *pool) {
+  const auto &type = data.type;
+  if (arrow::is_fixed_width(type->id())) {
+    buffer_sizes->push_back(0);
+    data_buffers->push_back(nullptr);
+    return Status::OK();
+  }
+
+  if (arrow::is_binary_like(type->id())) {
+    buffer_sizes->push_back((int) ((data.length + 1) * sizeof(int32_t)));
+    data_buffers->push_back(reinterpret_cast<const uint8_t*>(data.GetValues<int32_t>(1)));
+    return Status::OK();
+  }
+
+  if (arrow::is_large_binary_like(type->id())) {
+    buffer_sizes->push_back((int) ((data.length + 1) * sizeof(int64_t)));
+    data_buffers->push_back(reinterpret_cast<const uint8_t*>(data.GetValues<int64_t>(1)));
+    return Status::OK();
+  }
+
+  return {Code::Invalid, "unsupported offset type for serialization " + type->ToString()};
+}
+
+Status CylonTableSerializer::Make(const std::shared_ptr<Table> &table,
+                                  std::shared_ptr<TableSerializer> *serializer) {
```

Comment:
why this is not std::shared_ptr<TableSerializer> &serializer

Reply:
This is because of the google style guide. We initially followed it, but then decided to use mutable references for output params. I wanted to reverse it (again) mainly because of [this](https://stackoverflow.com/a/26441753/4116268) reason TBH.

Reply:
I kind of like using & in the method signature and avoid using & when calling. But I get the point.